tests/integration/auth_utils.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+from base64 import b64encode
+from os.path import join
+from secrets import token_urlsafe
+from urllib.parse import urlencode
+
+import requests
+
+
+def construct_authorization_url(hydra_url: str, client_id: str, client_secret: str) -> str:
+    params = {
+        "client_id": client_id,
+        "redirect_uri": client_secret,
+        "response_type": "code",
+        "response_mode": "query",
+        "scope": "openid profile email",
+        "state": token_urlsafe(),
+        "nonce": token_urlsafe(),
+    }
+    return join(hydra_url, "oauth2/auth?" + urlencode(params))
+
+
+def get_basic_http_auth_header(username: str, password: str) -> str:
+    token = b64encode(f"{username}:{password}".encode("utf-8")).decode("ascii")
+    return f"Basic {token}"
+
+
+def client_credentials_grant_request(
+    hydra_url: str, client_id: str, client_secret: str, scope: str = "openid profile"
+) -> requests.Response:
+    url = join(hydra_url, "oauth2/token")
```

Comment:
I prefer using `os.path.join`, as the urllib one can be a little unpredictable if there is no trailing slash. E.g.:
```console
>>> import os
>>> import urllib
>>> os.path.join("http://whatever.com/a/b", "c")
'http://whatever.com/a/b/c'
>>> os.path.join("http://whatever.com/a/b/", "c")
'http://whatever.com/a/b/c'
>>> os.path.join("http://whatever.com/a/b//", "c")
'http://whatever.com/a/b//c'
>>> urllib.parse.urljoin("http://whatever.com/a/b", "c")
'http://whatever.com/a/c'
>>> urllib.parse.urljoin("http://whatever.com/a/b/", "c")
'http://whatever.com/a/b/c'
>>> urllib.parse.urljoin("http://whatever.com/a//b/", "c")
'http://whatever.com/a/b/c'
```

On one hand we get a url with double `/` (which will still work), on the other hand we get a url which is missing part of the path. I would like to keep it this way, if that's ok.

Reply:
@nsklikas so what we are saying is that `http://abc.com/a//b/c` is something we consider valid?if yes no problem, but from a backend purpose it's not generally correct afaik....frontend wise i have absolutely no clue, especially with single page apps

Reply:
It is a valid URL, although `http://abc.com/a//b/c` != `http://abc.com/a/b/c`. I think that most servers would replace `//` with `/`, BUT if you input is `http://whatever.com/a//b/`, it makes more sense to me for the output to be `http://whatever.com/a//b/...` (contain the `//`) as well, rather than removing the `//` when the operation that I expect from the function is `join`.

I am not sure how browsers treat `//`, but I think that they make the request as is (with the `//`)